File: motio_client/lib/models/todo_list.dart
import 'todo_item.dart';
import 'user.dart';

class TodoList {
  final int id;
  final String listName;
  final List<TodoItem> items;
  final User createdByUser;
  final List<User> accessibleUsers;

  TodoList({
    required this.id,
    required this.listName,
    required this.items,
    required this.createdByUser,
    required this.accessibleUsers,
  });

  factory TodoList.fromJson(Map<String, dynamic> json) {
    return TodoList(
      id: json['id'],
      listName: json['listName'],
      items: (json['items'] as List).map((i) => TodoItem.fromJson(i)).toList(),
      createdByUser: User.fromJson(json['createdByUser']),
      accessibleUsers: (json['accessibleUsers'] as List).map((u) => User.fromJson(u)).toList(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'listName': listName,
      'items': items.map((i) => i.toJson()).toList(),
      'createdByUser': createdByUser.toJson(),
      'accessibleUsers': accessibleUsers.map((u) => u.toJson()).toList(),
    };
  }

  TodoList copyWith({
    int? id,
    String? listName,
    List<TodoItem>? items,
    User? createdByUser,
    List<User>? accessibleUsers,
  }) {
    return TodoList(
      id: id ?? this.id,
      listName: listName ?? this.listName,
      items: items ?? this.items,
      createdByUser: createdByUser ?? this.createdByUser,
      accessibleUsers: accessibleUsers ?? this.accessibleUsers,
    );
  }
}

File: motio-commons/src/main/java/com/motio/commons/model/TodoList.java
package com.motio.commons.model;

import com.fasterxml.jackson.annotation.JsonIdentityInfo;
import com.fasterxml.jackson.annotation.ObjectIdGenerators;
import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.util.HashSet;
import java.util.LinkedList;
import java.util.List;
import java.util.Set;

@Data
@NoArgsConstructor
@AllArgsConstructor
@Entity
@Table(name = "todo_list")
@JsonIdentityInfo(generator = ObjectIdGenerators.PropertyGenerator.class, property = "id")
public class TodoList {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false)
    private String listName;

    @OneToMany(cascade = CascadeType.ALL, orphanRemoval = true)
    private List<TodoItem> items = new LinkedList<>();

    @ManyToOne
    @JoinColumn(name = "created_by_user_id", nullable = false)
    private User createdByUser;

    @ManyToMany
    @JoinTable(
            name = "todo_list_access",
            joinColumns = @JoinColumn(name = "todo_list_id"),
            inverseJoinColumns = @JoinColumn(name = "user_id")
    )
    private Set<User> accessibleUsers = new HashSet<>();

    public TodoList createCopy() {
        TodoList todoList = new TodoList();
        todoList.setId(getId());
        todoList.setListName(getListName());
        todoList.setCreatedByUser(getCreatedByUser());
        todoList.setAccessibleUsers(new HashSet<>(getAccessibleUsers()));
        todoList.setItems(new LinkedList<>(getItems()));
        return todoList;
    }
}
